l10n_fi_business_code_validate/models/res_partner_id_category.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Oy Tawasta OS Technologies Ltd.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import re
+from odoo import models
+from odoo import _
+from odoo.exceptions import ValidationError
+
+
+class ResPartnerIdCategory(models.Model):
+    _inherit = 'res.partner.id_category'
+
+    # Business id validator
+    def validate_business_id(self, id_number):
+        if not id_number:
+            # Business id is not set. This is fine.
+            return True
+
+        partner = id_number.partner_id
+
+        if partner.country_id:
+            # Try to find a validator function by partner country code
+
+            # The method name should be
+            # '_business_id_validate_{lowercase_country_code}'
+            validator_method_name = '_business_id_validate_%s' % \
+                                    partner.country_id.code.lower()
+
+            # Check if the method exists
+            if hasattr(self, validator_method_name):
+                # Run the validator
+                validator_method = getattr(self, validator_method_name)
+                validator_method(partner)
+
+    # Finnish (FI) business id validation
+    def _business_id_validate_fi(self, partner):
+        # Update the format
+        self._business_id_update_format_fi(partner)
+
+        business_id = partner.business_id
+
+        if re.match('^[0-9]{3}[.][0-9]{3}$', business_id):
+            # Registered association (rekisteröity yhdistys, ry / r.y.).
+            # Format 123.456
+            return True
+
+        # Validate business id formal format
+        if not re.match('^[0-9]{7}[-][0-9]{1}$', business_id):
+            msg = _('Your business id is invalid. Please use format 1234567-1')
+            raise ValidationError(msg)
+
+        # The formal format is ok, check the validation number
+        multipliers = [7, 9, 10, 5, 8, 4,
+                       2]  # Number-space specific multipliers
```

Comment:
Could the comment be above, so the multipliers could fit on a single line?

Reply:
Also, if these are constant, they maybe should be in module-scope as a constant variable:
```
FINNISH_ID_DIGIT_MULTIPLIERS = [7, 9, 10, 5, 8, 4, 2]

class ResPartnerIdCategory(models.Model):
    _inherit = 'res.partner.id_category'
...
```
Feel free to come up with a better name.